### PR TITLE
Reland "update ScrollMetricsNotification (#87421)"

### PR DESCRIPTION
**Pull request by chunhtai**
This reverts commit 0f465e5b2a3ed2431321b490a614c3d15089854c.

previous pr broke internal tests. The test will be fixed by cl/388794349

blocked on cl/388794349
## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chunhtai** [comment]
cc @Piinks  @xu-baolin

**Comment by xu-baolin** [comment]
Some tests failed because of multiple assertions reported.

I am surprised that the test did not report an error when it first landing.

**Comment by chunhtai** [comment]
https://github.com/flutter/flutter/commit/7bbcdc239b changed the way the error is reported

**Review by Piinks — COMMENTED** [review]
It looks like the failing tests reveal a condition that should be added to the changes from 7bbcdc2
If more than one position is attached to the controller, then `RawScrollbarState._shouldUpdatePainter` should return false. The failing checks appear to be from the gallery integration tests.
@chunhtai do you want to add that here? Or should I add it in a separate change?

**Comment by chunhtai** [comment]
@Piinks I will add it in this pr

File: packages/flutter/lib/src/widgets/scroll_controller.dart [hunk 1]
```diff
@@ -88,7 +88,6 @@ class ScrollController extends ChangeNotifier {
   ///
   /// This should not be mutated directly. [ScrollPosition] objects can be added
   /// and removed using [attach] and [detach].
-  @protected
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_controller.dart:91]: Ooo this is a very exciting change. There are so many work arounds for this that we could clean up. :)

**Review by Piinks — APPROVED** [review]
LGTM!

**Comment by chunhtai** [comment]
Test run https://fusion2.corp.google.com/presubmit/tap/389684836/OCL:389684836:BASE:389709544:1628540380244:3438f481 

one golden failure that is a pretty minor diff and is probably the correct change merging this pr

**Pull request merged**